Sming/Core/Data/Stream/SectionTemplate.h

```diff
@@ -0,0 +1,179 @@
+ * http://github.com/SmingHub/Sming
+ * All files of the Sming Core are provided under the LGPL v3 license.
+ *
+ * SectionTemplate.h
+ *
+ * @author mikee47 <mike@sillyhouse.net> Nov 2020
+ *
+ ****/
+
+#pragma once
+
+#include "SectionStream.h"
+#include "TemplateStream.h"
+#include <Data/BitSet.h>
+#include "../Format.h"
+
+/**
+ * @brief Command map
+ *
+ * Commands may have zero or more arguments, which may be:
+ * 		Number:		starts with a digit
+ * 		String:		"..."
+ * 		expression: {...}
+ *
+ * Anything else is treated as a variable name.
+ * Separator is :
+ *
+ * @note Command tags are prefixed with _ to allow use of reserved words
+ * in the Command enumeration. This represents the ! prefix in actual use.
+ */
+#define SECTION_TEMPLATE_COMMAND_MAP(XX)                                                                               \
+	XX(_as_int, "{!int:A} Output A as integer")                                                                        \
+	XX(_as_float, "{!float:A} Output A as float")                                                                      \
+	XX(_as_string, "{!string:A} Output A as quoted string")                                                            \
+	XX(_mime_type, "{!mime_type:A} Get MIME type string for a filename")                                               \
+	XX(_replace, "{!replace:A:B:C} Copy of A with all occurrences of B replaced with C")                               \
+	XX(_length, "{!length:A} Number of characters in A")                                                               \
+	XX(_pad,                                                                                                           \
+	   "{!pad:A:B:C} Copy of A padded to at least B characters with C (default is space). Use -ve B to left-pad. C")   \
+	XX(_repeat, "{!repeat:A:B} Repeat A, number of iterations is B")                                                   \
+	XX(_kb, "{!kb:A} Convert A to KB")                                                                                 \
+	XX(_ifdef, "{!ifdef:A}block{/if} emit block if A is not zero-length")                                              \
+	XX(_ifndef, "{!ifdef:A}block{/if} emit block if A is zero-length")                                                 \
+	XX(_ifeq, "{!ifeq:A:B} emit block if A == B")                                                                      \
+	XX(_ifneq, "{!ifneq:A:B} emit block if A != B")                                                                    \
+	XX(_ifgt, "{!ifgt:A:B} emit block if A > B")                                                                       \
+	XX(_iflt, "{!iflt:A:B} emit block if A < B")                                                                       \
+	XX(_ifge, "{!ifge:A:B} emit block if A >= B")                                                                      \
+	XX(_ifle, "{!ifle:A:B} emit block if A <= B")                                                                      \
```

Comment:
Wow, this "simple" templating mechanism starts to get more and more powerful and will one day beat PHP :)

Reply:
This kind of evolved but is there a standard templating language that we should be using?

Reply:
> This kind of evolved but is there a standard templating language that we should be using?

Hm... this is a very good question. It is better, adoption-wise, to use something that is already easy and popular.  I can suggest [Twig](https://twig.symfony.com/) used in the PHP world,  [Jinja for C++](https://github.com/hughperkins/Jinja2CppLight) or [Nunjucks](https://mozilla.github.io/nunjucks/). There are plenty of templating engines out there and I am open for suggestions.

Reply:
On reflection I'm inclined to leave this 'engine' as-is.

The main reason I wrote this initially was to find a generic way to produce a directory listing in JSON format. I initially used a custom stream class, but then I realised templating could offer a more flexible solution. The browser can then do all the formatting, etc. to produce reports. However, I got carried away of course and the conditional stuff got added!

So the use cases are rather different I think. It looks like Jinja or Nunjucks could be added if the need arises for those more advanced features. They seem fairly consistent in using the {% ... %} escaping, however the command syntax here is much simpler so I think keeping {! ...} helps to differentiate it.

Reply:
> On reflection I'm inclined to leave this 'engine' as-is.

Agreed, we can always improve it if there is need for it.